### ISSUE TITLE
RuntimeError: Invalid quantization params for op RESHAPE at index 36 in subgraph 0

### ISSUE THREAD

**Issue by conglanjun** [tensorflow/tensorflow#35625]
I download source code of SSD from github.com. I want to convert the model to run in edgetpu.
When I convert saved model to tflite, then run 'edgetpu_compiler XXX.tflite' for running in tpu. But it tell me that not quantization, I change parameter for quantization.Such as 
converter.target_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8], 
but occur error: RuntimeError: Invalid quantization params for op RESHAPE at index 36 in subgraph 0

**System information**

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04.6 LTS
- TensorFlow installed from (source or binary): binary
- TensorFlow version (or github SHA if from source):1.15

**Command used to run the converter or code if you’re using the Python API**

```
import tensorflow as tf
from PIL import Image
import numpy as np

ckpt_filename = '/home/lanjunc/python_project/SSD-Tensorflow/convert/model'

img = Image.open('./dog.jpg')
img = img.resize((300, 300))
input_data = np.array(img, dtype=np.float32)
def representative_dataset_gen():
    for i in range(1):
        yield [input_data]


converter = tf.lite.TFLiteConverter.from_saved_model(ckpt_filename,
                                                     # tag_set=['test_saved_model1'], signature_key='signature_test1',
                                                     input_shapes={"Placeholder":[300,300,3]})

converter.optimizations = [tf.lite.Optimize.DEFAULT]  # DEFAULT, OPTIMIZE_FOR_SIZE, OPTIMIZE_FOR_LATENCY

converter.target_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8, tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
# converter.target_spec.supported_ops = [
#                                        tf.lite.OpsSet.SELECT_TF_OPS]
#
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
converter.representative_dataset = representative_dataset_gen

tflite_model = converter.convert()

open("converted_model.tflite", "wb").write(tflite_model)
```

**The output from the converter invocation**

```
RuntimeError: Invalid quantization params for op RESHAPE at index 36 in subgraph 0
```

**Comment by conglanjun** [comment]
I change my code

img1 = Image.open('./dog.jpg')
    img1 = img1.resize((300, 300))
    input_data = np.array(img1, dtype=np.float32)


    def representative_dataset_gen():
        for i in range(1):
            yield [input_data]

    # tf.contrib.quantize.create_eval_graph(isess.graph)
    toco_converter = tf.lite.TFLiteConverter.from_session(isess, input_tensors=[image_pre], output_tensors=localisations)
    toco_converter.optimizations = [tf.lite.Optimize.DEFAULT]
    toco_converter.target_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8, tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
    toco_converter.post_training_quantize = False
    toco_converter.inference_type = tf.uint8
    toco_converter.inference_input_type = tf.uint8
    toco_converter.quantized_input_stats = {'ssd_preprocessing_train/resize_image/Reshape': (0., 255.)}
    toco_converter.representative_dataset = representative_dataset_gen
    model_tflite_binary = toco_converter.convert()
    open("converted_model.tflite", "wb").write(model_tflite_binary)

-------------------------------------------
but occur the error:

Array ssd_300_vgg/conv1/conv1_1/Relu, which is an input to the Conv operator producing the output array ssd_300_vgg/conv1/conv1_2/Relu, is lacking min/max data, which is necessary for quantization. If accuracy matters, either target a non-quantized output format, or run quantized training with your model from a floating point checkpoint to change the input graph to contain min/max information. If you don't care about accuracy, you can pass --default_ranges_min= and --default_ranges_max= for easy experimentation.

**Comment by conglanjun** [comment]
I saw it:
https://medium.com/tensorflow/tensorflow-model-optimization-toolkit-post-training-integer-quantization-b4964a1ea9ba
It says that To execute entirely on specialized hardware that does not support floating point operations at all (for example, some machine learning accelerators, including the Edge TPU), you can specify a flag in order to output only integer operations:

When this flag is used and an operation has no integer quantizable counterpart, the TensorFlow Lite Converter will throw an error.
I want to run the model in edge tpu.I know the tpu only recognizes quantion model.If the model has the ops which can't be quantized, the model can't run in the tpu. Yes?

**Comment by conglanjun** [comment]
I compiler source code , and I try

bazel run --config=opt tensorflow/lite/toco:toco -- --input_file=/home/lanjunc/python_project/SSD-Tensorflow/convert/pb/ssd_frozen_model.pb --output_file=/home/lanjunc/python_project/SSD-Tensorflow/convert/ssd_tpu_compiler.tflite --input_shapes=300,300,3 --input_arrays=Placeholder --output_arrays='ssd_300_vgg/softmax/Reshape_1,ssd_300_vgg/softmax_1/Reshape_1,ssd_300_vgg/softmax_2/Reshape_1,ssd_300_vgg/softmax_3/Reshape_1,ssd_300_vgg/softmax_4/Reshape_1,ssd_300_vgg/softmax_5/Reshape_1,ssd_300_vgg/block4_box/Reshape,ssd_300_vgg/block7_box/Reshape,ssd_300_vgg/block8_box/Reshape,ssd_300_vgg/block9_box/Reshape,ssd_300_vgg/block10_box/Reshape,ssd_300_vgg/block11_box/Reshape,ssd_preprocessing_train/strided_slice' --inference_type=QUANTIZED_UINT8 --mean_values=128 --std_values=128 --default_ranges_min=0 --default_ranges_max=255

occur error:

2020-01-07 20:31:16.292889: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 254 operators, 369 arrays (0 quantized)
2020-01-07 20:31:16.295754: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 254 operators, 369 arrays (0 quantized)
2020-01-07 20:31:16.364710: I tensorflow/lite/toco/graph_transformations/identify_dilated_conv.cc:202] Replaced sub-network with Dilated Conv2D op outputting "ssd_300_vgg/conv6/Conv2D".
2020-01-07 20:31:16.437956: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 78 operators, 176 arrays (0 quantized)
2020-01-07 20:31:16.439118: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 2: 73 operators, 170 arrays (0 quantized)
2020-01-07 20:31:16.439996: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 3: 72 operators, 168 arrays (0 quantized)
2020-01-07 20:31:16.440905: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before pre-quantization graph transformations: 72 operators, 168 arrays (0 quantized)
2020-01-07 20:31:16.441421: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before Group bidirectional sequence lstm/rnn: 72 operators, 168 arrays (0 quantized)
2020-01-07 20:31:16.441973: W tensorflow/lite/toco/tooling_util.cc:1288] Fixing constant output array ssd_preprocessing_train/strided_slice by inserting a copy. This is not optimal.
2020-01-07 20:31:16.442080: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before default min-max range propagation graph transformations: 73 operators, 170 arrays (0 quantized)
2020-01-07 20:31:16.442621: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After default min-max range propagation graph transformations pass 1: 73 operators, 170 arrays (0 quantized)
2020-01-07 20:31:16.443268: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before quantization graph transformations: 73 operators, 170 arrays (0 quantized)
Unimplemented: this graph contains an operator of type Cast for which the quantized form is not yet implemented. Sorry, and patches welcome (that's a relatively fun patch to write, mostly providing the actual quantized arithmetic code for this op).

**Comment by jvishnuvardhan** [comment]
@conglanjun Can you please share a standalone code to reproduce your issue? Thanks!

**Comment by conglanjun** [comment]
I saw 
https://github.com/tensorflow/tensorflow/issues/20955

```
suharshs commented on Aug 18, 2018

Regarding cast, why does the model have a Cast operation?
Make sure you are converting a eval graph and not a train graph which can sometimes have spurious unsupported operations.

Are you trying to quantize the model using the tool here: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/quantize

you should train with that which should place quantization ops in the graph to collect info. That second error you get means that the graph is not getting the correct quantization operations in the graph.

How are you making your fake quantized frozen graph to provide to tflite conversion?
```
but https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/quantize 404 not found
tf 2.0 remove the pah?

```
raninbowlalala commented on Aug 20, 2018
@suharshs
For cast op, I want to use dummy quantization for deeplabv3+ (mobilenetv2) model named "mobilenetv2_coco_voc_trainaug" which download from https://github.com/tensorflow/models/blob/master/research/deeplab/g3doc/model_zoo.md. I use tensorboard to see the graph and there are "cast" op.

For training a quantization model. I am sorry I forgot to add "tf.contrib.quantize.create_eval_graph()", convert model successfully after I add this command. Thank you very much!
```
I should add "tf.contrib.quantize.create_eval_graph()"?

**Comment by conglanjun** [comment]
I saw
https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/contrib/quantize/README.md
I should train my model, using "Quantization-aware training".
And make pb file using create_eval_graph().

And I use the SSD code is 
https://github.com/balancap/SSD-Tensorflow
I want to use this model to run in EdgeTpu.

**Comment by conglanjun** [comment]
I update new message:
I use tf.contrib.quantize.create_training_graph() in training, save to ckpt
I load ckpt, then use tf.contrib.quantize.create_eval_graph(), save to saved model.

I use the saved model to convert to tflite. Code is:
```
import tensorflow as tf
from PIL import Image
import numpy as np

ckpt_filename = '/home/lanjunc/python_project/SSD-Tensorflow/convert/model'
# ckpt_filename = '../notebooks/ssd_300_vgg_test'
img = Image.open('./dog.jpg')
img = img.resize((300, 300))
input_data = np.array(img, dtype=np.float32)
def representative_dataset_gen():
    for i in range(1):
        yield [np.random.randn(300, 300, 3)]


converter = tf.lite.TFLiteConverter.from_saved_model(ckpt_filename, input_shapes={"input":[300,300,3]})

converter.optimizations = [tf.lite.Optimize.DEFAULT]  # DEFAULT, OPTIMIZE_FOR_SIZE, OPTIMIZE_FOR_LATENCY

converter.target_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8, tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
converter.representative_dataset = representative_dataset_gen

tflite_model = converter.convert()

open("converted_model.tflite", "wb").write(tflite_model)
```
`2020-01-08 17:05:57.434156: F tensorflow/lite/toco/import_tensorflow.cc:2471] Check failed: status.ok() Neither input_content (0) nor float_val (294911) have the right dimensions (294912) for this float tensor
	 (while processing node 'ssd_300_vgg/conv3/conv3_1/weights_quant/FakeQuantWithMinMaxVars')
Aborted (core dumped)`

I think input shape is error?
Quantization result input is "dimensions (294912)"?
But the input shape is [300,300,3]. never changed!

**Comment by conglanjun** [comment]
I saw the source code https://git.codingcafe.org/Mirrors/tensorflow/tensorflow/blob/f4cb5978667ccf6396e4a779e3a482766959e5dd/tensorflow/contrib/lite/toco/import_tensorflow.cc
```
    return Status(
        false,
        absl::StrCat("Neither input_content (",
                     input_tensor.tensor_content().size() / sizeof(float),
                     ") nor float_val (", input_tensor.float_val_size(),
                     ") have the right dimensions (", input_flat_size,
                     ") for this float tensor"));

```
input_flat_size != input_tensor.float_val_size()
how can I control the size?

**Comment by jvishnuvardhan** [comment]
@conglanjun Can you please share a standalone code to reproduce your issue? Thanks!

**Comment by conglanjun** [comment]
https://drive.google.com/open?id=18IoP7I_TcixoJtje2vDsbFtpOYDTl-Fv
you can download it.

**Comment by conglanjun** [comment]
I tell you the code.
Training the model, use the script, SSD-Tensorflow/train_ssd_network.py
run
```
DATASET_DIR=./tfrecords
TRAIN_DIR=./logs/
CHECKPOINT_PATH=./checkpoints/ssd_300_vgg.ckpt
python train_ssd_network.py \
    --train_dir=${TRAIN_DIR} \
    --dataset_dir=${DATASET_DIR} \
    --dataset_name=pascalvoc_2007 \
    --dataset_split_name=train \
    --model_name=ssd_300_vgg \
    --checkpoint_path=${CHECKPOINT_PATH} \
    --save_summaries_secs=60 \
    --save_interval_secs=600 \
    --weight_decay=0.0005 \
    --optimizer=adam \
    --learning_rate=0.001 \
    --batch_size=32

```
I add the code
`
g = tf.get_default_graph()
tf.contrib.quantize.create_training_graph(input_graph=g, quant_delay=200)
`
in SSD-Tensorflow/train_ssd_network.py.
It will product model such as 'SSD-Tensorflow/logs/model.ckpt-5'
Then I run SSD-Tensorflow/notebooks/ssd_notebookv_eval.py
I have add 'tf.contrib.quantize.create_eval_graph()' in it.
I'll create saved model in SSD-Tensorflow/notebooks/model.
Then I want to create tflite.
Run 'SSD-Tensorflow/convert/tf_lite_convert.py'
occur the error:
RuntimeError: Invalid quantization params for op RESHAPE at index 80 in subgraph 0.
I download tensorflow source code, want to chang tensorflow code for fixing the error.
I change the file:
'tensorflow/tensorflow/tensorflow/lite/tools/optimize/quantize_model.cc'
```
} else if (tensor_property.restriction) {
    const auto scale_and_zp = tensor_property.restricted_value;
    // Apply to output.
    output_tensor->quantization = absl::make_unique<QuantizationParametersT>();
    output_tensor->quantization->scale.push_back(scale_and_zp.first);
    output_tensor->quantization->zero_point.push_back(scale_and_zp.second);
    output_tensor->type = TensorType_INT8;
}
```
as
```
} else if (tensor_property.restriction) {
    const auto scale_and_zp = tensor_property.restricted_value;
    // Apply to output.
    output_tensor->quantization = absl::make_unique<QuantizationParametersT>();
    output_tensor->quantization->scale.push_back(scale_and_zp.first);
    output_tensor->quantization->zero_point.push_back(scale_and_zp.second);
    output_tensor->type = TensorType_INT8;
    // patch: read min/max from the input tensor
    const int input_tensor_idx = op->inputs[property.inputs[0].first];
    TensorT* input_tensor = subgraph->tensors[input_tensor_idx].get();
    output_tensor->quantization->min.push_back(input_tensor->quantization->min[0]);
    output_tensor->quantization->max.push_back(input_tensor->quantization->max[0]);
}
```
Why can come in the case:'else if (tensor_property.restriction) {'?
I add max and min in it.
But other error:
```
error_reporter->Report(
          "Invalid quantization params for op %s at index %d "
          "in subgraph %d, scale.size:%d, zero_point.size:%d, min.size:%d, max.size:%d",
          EnumNameBuiltinOperator(op_code), op_idx, subgraph_idx, input_tensor->quantization->scale.size(), input_tensor->quantization->zero_point.size(), input_tensor->quantization->min.size(), input_tensor->quantization->max.size());
      
```
I printf it. show that min.size()==0,max.size()==0
I don't know why there isn't max,min in input_tensor.
when Quantizing train model, and eval model occur the error?
I set min and max default value is [-127, 128].
```
if(input_tensor->quantization->min.size() == 0 || input_tensor->quantization->max.size() == 0){
        input_tensor->quantization->min.push_back(-128);
        input_tensor->quantization->max.push_back(127);
}
```
Other error:RuntimeError: Unsupported input type UINT8 for input tensor 0 of type FLOAT32
how my model have float32? I have qunitized.
I find the source code.
```
for (int i = 0; i < subgraph->inputs.size(); ++i) {
      TensorT* tensor = subgraph->tensors[subgraph->inputs[i]].get();
      // TODO(suharshs): Add support for this case if it ever comes up.
      if (tensor->type == TensorType_FLOAT32 && input_type != tensor->type) {
        error_reporter->Report(
            "Unsupported input type %s for input tensor %d of type %s.",
            EnumNameTensorType(input_type), subgraph->inputs[i],
            EnumNameTensorType(tensor->type));
        return kTfLiteError;
      }
      const int32_t input_idx =
          SetInputType(model, subgraph, subgraph->inputs[i], input_type);
      if (input_idx < 0) {
        continue;
      }
      subgraph->inputs[i] = input_idx;
    }
```
It show that // TODO(suharshs): Add support for this case if it ever comes up.
Maybe, the process which I quantize the model, is error!
Can help me?
I want to train my own model, quantize it,product tflite, runing the model in EgeTpu.

**Comment by skavulya** [comment]
I had a similar problem with tflite post-training quantization for the SSD + mobilenetv3 model on TensorFlow 1.15.0. I resolved the issue by upgrading to TensorFlow 2.1.0 then saving the model as v1 compatible.

```
import tensorflow.compat.v1 as tf

converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
tflite_quant_model = converter.convert()
```

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
